sdk/eventhub/azure-eventhubs/azure/eventhub/receiver.py

```diff
@@ -285,45 +290,53 @@ def receive(self, max_batch_size=None, timeout=None):
                 :caption: Receive events from the EventHub.
 
         """
-        if self.error:
-            raise self.error
-        if not self.running:
-            raise ValueError("Unable to receive until client has been started.")
-        data_batch = []
-        try:
-            timeout_ms = 1000 * timeout if timeout else 0
-            message_batch = self._handler.receive_message_batch(
-                max_batch_size=max_batch_size,
-                timeout=timeout_ms)
-            for message in message_batch:
-                event_data = EventData(message=message)
-                self.offset = event_data.offset
-                data_batch.append(event_data)
-            return data_batch
-        except (errors.TokenExpired, errors.AuthenticationException):
-            log.info("Receiver disconnected due to token error. Attempting reconnect.")
-            self.reconnect()
-            return data_batch
-        except (errors.LinkDetach, errors.ConnectionClose) as shutdown:
-            if shutdown.action.retry and self.auto_reconnect:
-                log.info("Receiver detached. Attempting reconnect.")
-                self.reconnect()
+        time_out_at = time.time() + (timeout if timeout else 0)
+        num_tries = 0 # If timeout is 0, have to make it run at least once.
+        while num_tries == 0 \
+                or ((max_reconnect_retries is None or num_tries <= max_reconnect_retries) \
+                and (timeout is None or time.time() < time_out_at)):
```

Comment:
Do we want to raise some kind of timeout error in this case?

Reply:
Ah - I guess this is to reproduce the current timeout behaviour where receive simply returns after the timeout?

Reply:
Ish.  We could get away with throwing a timeout exception, because I've decided to be paranoid and make the default params such that it'll be fully backcompat.  I'm fine with adding that if we want to be principled. (Candidly I mostly hadn't because this was primarily just a compliment to adding max_retry; I always pair the two concepts when setting up circuit breakers, and yes, as you say in the existing behavior it just tries once then silently falls through)

Reply:
I would prefer back compat in this case I think.
Otherwise I guess it's potentially a breaking behaviour change?

Reply:
Nevermind - I see this is a new opt-in feature - so non breaking :)